### PR TITLE
Added container instance creation and takedown

### PR DESCRIPTION
**Pull request by schmerl**

### CODE REVIEW

File: src/rosdiscover/cli.py [hunk 1]
```diff
@@ -182,11 +181,22 @@ def _periodic_observe(interval: float, args: argparse.Namespace) -> SystemSummar
     return summary
 
 
-def observe(args) -> None:
+def do_observe(obs, args):
```
Comment (ChrisTimperley) [inline src/rosdiscover/cli.py:184]: type annotations would be helpful for readability here

File: src/rosdiscover/cli.py [hunk 2]
```diff
@@ -110,18 +110,17 @@ def generate_acme(args) -> None:
         acme_gen.check_acme()
 
 
-def _observe(args) -> SystemSummary:
-    config = Config.from_yaml_string(args.config)
-    obs = Observer.for_container(args.container, config)
+def _observe(obs, args) -> SystemSummary:
```
Comment (ChrisTimperley) [inline src/rosdiscover/cli.py:113]: same comment about type annotations

File: src/rosdiscover/cli.py [hunk 3]
```diff
@@ -110,18 +110,17 @@ def generate_acme(args) -> None:
...
-def _periodic_observe(interval: float, args: argparse.Namespace) -> SystemSummary:
-    config = Config.from_yaml_string(args.config)
-    obs = Observer.for_container(args.container, config)
+def _periodic_observe(obs: Observer,
+                      interval: float,
+                      args: argparse.Namespace) -> SystemSummary:
+
     launches: t.Sequence[Popen] = []
     if args.do_launch:
         launches = obs.launch_from_config(args.launch_sleep)
```
Comment (ChrisTimperley) [inline src/rosdiscover/cli.py:126]: launches doesn't seem to be used?

File: src/rosdiscover/observer/observer.py [hunk 4]
```diff
@@ -21,6 +23,7 @@ class ExecutionError(Exception):
 
 
 class Observer(ABC):
+    _attached: t.Optional[Popen]
```
Comment (ChrisTimperley) [inline src/rosdiscover/observer/observer.py:26]: what is `_attached`?

File: src/rosdiscover/observer/observer.py [hunk 5]
```diff
@@ -45,17 +48,69 @@ def for_container(cls,
         rsw = roswire.ROSWire()
         app = rsw.app(config.image, config.sources)
         instance = app.attach(container, require_description=True)
+        observer = cls._build_observer(app, config, instance)
+        observer._attached = None
+        return observer
+
+    @classmethod
+    def _build_observer(cls, app, config, instance):
```
Comment (ChrisTimperley) [inline src/rosdiscover/observer/observer.py:56]: type annotations

File: src/rosdiscover/cli.py (hunk 3 above)
Comment (schmerl) [reply src/rosdiscover/cli.py:126]: It's used in the finally block

File: src/rosdiscover/observer/observer.py [hunk 6]
```diff
@@ -45,17 +48,69 @@ def for_container(cls,
...
+            container.
+        """
+        rsw = roswire.ROSWire()
+        app = rsw.app(config.image, config.sources)
+        instance = app.launch(environment=config.environment)
+        observer = cls._build_observer(app, config, instance)
+        observer._attached = False
+
+        logger.debug("Starting the container")
+        cmd = "/startup-vnc.sh"
```
Comment (ChrisTimperley) [inline src/rosdiscover/observer/observer.py:88]: this seems hardcoded to TheRobotCooperative images? We should probably make this a parameter or create an issue.

File: src/rosdiscover/observer/observer.py [hunk 7]
```diff
@@ -45,17 +48,69 @@ def for_container(cls,
...
+        -------
+        Observer
+            An observer that is appropriate for the kind of ROS system that is running in the
+            container.
+        """
+        rsw = roswire.ROSWire()
+        app = rsw.app(config.image, config.sources)
+        instance = app.launch(environment=config.environment)
+        observer = cls._build_observer(app, config, instance)
+        observer._attached = False
```
Comment (ChrisTimperley) [inline src/rosdiscover/observer/observer.py:85]: this doesn't type check (`observer._attached` is a `t.Optional[Popen]`)

Comment (schmerl) [reply src/rosdiscover/observer/observer.py:85]: Funny how these things get through mypy

File: src/rosdiscover/observer/observer.py [hunk 8]
```diff
@@ -45,17 +48,78 @@ def for_container(cls,
...
+        config: Config,
+        instance: AppInstance):
         if app.description.distribution.ros == ROSVersion.ROS1:
             from .ros1 import ROS1Observer
             return ROS1Observer(instance, config)
         else:
             from .ros2 import ROS2Observer
             return ROS2Observer(instance, config)
 
+    @classmethod
```
Comment (ChrisTimperley) [inline src/rosdiscover/observer/observer.py:71]: I would suggest turning this method into a contextmanager and then creating and killing the container inside this method. That way, Observer doesn't need `._x_start_script` or `._own_instance`.

File: src/rosdiscover/observer/observer.py [hunk 9]
```diff
@@ -45,17 +48,78 @@ def for_container(cls,
...
+        """Constructs an observer by starting an instance of an image
+
+        Parameters
+        ----------
+        config: Config
+            A description of the configuration used by the running container.
+            Will contain the image name, sources, and environment variables to
+            run
+        start_script: str
+            An optional script (on the container) to run after instance cretion
```
Comment (ChrisTimperley) [inline src/rosdiscover/observer/observer.py:84]: missing newline

File: src/rosdiscover/observer/observer.py [hunk 10]
```diff
@@ -45,17 +48,78 @@ def for_container(cls,
...
+        Returns
+        -------
+        Observer
+            An observer that is appropriate for the kind of ROS system that is running in the
+            container.
+        """
+        rsw = roswire.ROSWire()
+        app = rsw.app(config.image, config.sources)
+        instance = app.launch(environment=config.environment)
```
Comment (ChrisTimperley) [inline src/rosdiscover/observer/observer.py:93]: `with` can be used here

File: src/rosdiscover/observer/observer.py [hunk 11]
```diff
@@ -45,13 +47,63 @@ def for_container(cls,
...
+                    cmd = start_script
+                    script = instance.shell.popen(cmd)
+                    time.sleep(5)
+                    if script.returncode and script.returncode != 1:
+                        for line in script.stream:
+                            logger.error(line)
+                        raise ExecutionError("Could not run start script")
+                yield observer
+            finally:
+                if start_script and script:
```
Comment (ChrisTimperley) [inline src/rosdiscover/observer/observer.py:104]: why `and script`?

Comment (schmerl) [reply src/rosdiscover/observer/observer.py:104]: Technically, script could be unassigned if `logger.debug` throws an error.

**Pull request merged**